keps/sig-scheduling/20190221-even-pods-spreading.md

```diff
@@ -0,0 +1,261 @@
+
+### User Stories
+
+#### Story 1 - NodeAffinity
+
+As an application developer, I want my application pods to be scheduled onto
+specific topology domains (via NodeAffinity with spec "app in [zone1, zone2,
+zone3]"), but I don't want them to be stacked too much on one topology. (see
+[#68981](https://github.com/kubernetes/kubernetes/issues/68981))
+
+#### Story 2 - PodAffinity
+
+As an application developer, I want my application pods to co-exist with
+particular pods in the same topology domain (via PodAffnity), and I want them to
+be deployed onto separate nodes as even as possible.
+
+#### Story 3 - PodAntiAffinity
+
+As an application developer, I want my application pods not to co-exist with
+specific pods (via PodAntiAffinity). But in some cases it'd be favorable to
+tolerate "violating" pods in a manageable way. For example, suppose an app
+(replicas=2) is using PodAntiAffinity and deployed onto a 2-nodes cluster, and
+next the app needs to perform a rolling upgrade, then a third replacement pod is
+created, but it failed to be placed due to lack of resource. In this case,
+
+- if HPA is enabled, a new machine will be provisioned to hold the new pod
+  (although old replicas will be deleted afterwards) (see
+  [#40358](https://github.com/kubernetes/kubernetes/issues/40358))
+- if HPA is not enabled, it's a deadlock since the replacement pod can't be
+  placed. The only workaround at this moment is to update app strategyType from
+  "RollingUpdate" to "Recreate".
+
+Both are not ideal solutions. A promising solution is to give user an option to
+trigger "toleration" mode when the cluster is out of resource. Then in
+aforementioned example, a third pod is "tolerated" to be put onto node1 (or
+node2). But keep it in mind, this behavior is only triggered upon resource
+shortage. For a 3-nodes cluster, the third pod will still be placed onto node3
+(if node3 is capable).
+
+### Risks and Mitigations
+
+Along with this feature, inevitable cost will be applied each time on
+scheduling. So to mitigate potential performance impact, initial implementation
+will limit the semantics of "even spreading" on `kubernetes.io/hostname` for
+PodAffinity and PodAntiAffinity.
+
+## Design Details
+
+We'd like to propose a new structure called `EvenSpreading`, which is a sub
+field of NodeAffinity, PodAffinity and PodAntiAffinity:
+
+```go
+type NodeAffinity struct {
+    EvenSpreading *EvenSpreading
+    ......
+}
+
+type PodAffinity struct {
+    EvenSpreading *EvenSpreading
+    ......
+}
+
+type PodAntiAffinity struct {
+    EvenSpreading *EvenSpreading
+    ......
+}
+```
+
+And it's only effective when (1) it's not nil and (2) "hard" affinity
+requirements (i.e.  `RequiredDuringSchedulingIgnoredDuringExecution`) are
+defined)
+
+API of `EvenSpreading` is defined as below:
+
+```go
+type EvenSpreading struct {
+    // MaxSkew describes the degree of imbalance of pods spreading.
+    // Default value is 1 and 0 is not allowed.
+    MaxSkew int32
```

Comment:
Setting a proper value to `masSkew` is still tricky in my mind ... any idea of practice/plan of this?

Reply:
One potential issue is, if a pod scheduling failed due to a spreading predicate, users tend to be confused about what should the proper `maxSkew` value unless they try for more times. I'm thinking how we can make it more straightforward to know, for example, `maxSkew=1` really means before schedule.

Reply:
> Setting a proper value to `maxSkew` is still tricky

I'd say default value 1 applies to most of the cases. It means for an incoming pod, if the existing pods spread as 2/1, the only chance you can get is 2/2 - 3/1 is not allowed. And even in the beginning I proposed the `maxSkew=1` design. But later on, I'd like to make it as a flexible option to adapt to more cases (probably :) let's see other folks' options)

> if a pod scheduling failed due to a spreading predicate, users tend to be confused

I'm less concerned b/c (1) this is an advanced feature which is not enabled by default. In other words, people who enable this mush have read related documentations on how it can impact their workloads, and (2) we will give descriptive predicates failure events so as not to confuse people. (in `kubectl describe pod xyz`)

Reply:
Agreed on "this is an advanced feature which is not enabled by default". 

It's reasonable to assume users of this feature should know what they are doing and have read the docs for many times. /cc @Fei-Guo

Reply:
Any specific reason for having this hard requirement(a predicate) instead of soft requirement(priority)?

Reply:
Major reason is MaxSkew provides a certain extent of "preference". So if it works in a hard requirement, it's sort of hard+soft; while if it works with a soft requirement, it's soft+soft. IMO it doesn't quite make sense as things are even more unpredictable, hence will be less useful.

And note that the above discussion thread is all based on Design A, which is mixed in Affinity spec.